File: services/indexer/pkg/store/postgres.go
package store

import (
	"context"
	"fmt"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/make-dot-tech/aefi/services/indexer/pkg/models"
)

// Store persists canonical events and the indexer cursor.
type Store interface {
	GetCursor(ctx context.Context, chainID int64) (models.Cursor, error)
	UpsertBatch(ctx context.Context, chainID int64, events []any, cursor models.Cursor) error
	Close()
}

type Postgres struct {
	pool *pgxpool.Pool
}

func NewPostgres(ctx context.Context, databaseURL string) (*Postgres, error) {
	pool, err := pgxpool.New(ctx, databaseURL)
	if err != nil {
		return nil, fmt.Errorf("pgx pool: %w", err)
	}
	if err := pool.Ping(ctx); err != nil {
		pool.Close()
		return nil, fmt.Errorf("pg ping: %w", err)
	}
	return &Postgres{pool: pool}, nil
}

func (p *Postgres) Close() {
	p.pool.Close()
}

func (p *Postgres) GetCursor(ctx context.Context, chainID int64) (models.Cursor, error) {
	var c models.Cursor
	err := p.pool.QueryRow(ctx, `
		SELECT chain_id, last_block, last_log_index
		FROM indexer_cursor WHERE chain_id = $1
	`, chainID).Scan(&c.ChainID, &c.LastBlock, &c.LastLogIndex)
	if err == pgx.ErrNoRows {
		return models.Cursor{ChainID: chainID}, nil
	}
	if err != nil {
		return c, err
	}
	return c, nil
}

func (p *Postgres) UpsertBatch(ctx context.Context, chainID int64, events []any, cursor models.Cursor) error {
	tx, err := p.pool.Begin(ctx)
	if err != nil {
		return err
	}
	defer tx.Rollback(ctx)

	for _, ev := range events {
		if err := upsertEvent(ctx, tx, ev); err != nil {
			return err
		}
	}

	_, err = tx.Exec(ctx, `
		INSERT INTO indexer_cursor (chain_id, last_block, last_log_index, updated_at)
		VALUES ($1,$2,$3, now())
		ON CONFLICT (chain_id) DO UPDATE SET
		  last_block = EXCLUDED.last_block,
		  last_log_index = EXCLUDED.last_log_index,
		  updated_at = now()
	`, chainID, cursor.LastBlock, cursor.LastLogIndex)
	if err != nil {
		return err
	}
	return tx.Commit(ctx)
}

func upsertEvent(ctx context.Context, tx pgx.Tx, ev any) error {
	switch e := ev.(type) {
	case models.TransferEvent:
		if err := upsertBase(ctx, tx, e.Base); err != nil {
			return err
		}
		_, err := tx.Exec(ctx, `
			INSERT INTO evt_transfer (id, from_addr, to_addr, value, decimals, emitter_role)
			VALUES ($1,$2,$3,$4,$5,$6)
			ON CONFLICT (id) DO UPDATE SET
			  from_addr = EXCLUDED.from_addr,
			  to_addr = EXCLUDED.to_addr,
			  value = EXCLUDED.value,
			  decimals = EXCLUDED.decimals,
			  emitter_role = EXCLUDED.emitter_role
		`, e.Base.ID, e.From, e.To, e.Value.String(), e.Decimals, e.EmitterRole)
		return err

	case models.MemoEvent:
		if err := upsertBase(ctx, tx, e.Base); err != nil {
			return err
		}
		_, err := tx.Exec(ctx, `
			INSERT INTO evt_memo (id, sender, memo_id, payload, call_data_hash)
			VALUES ($1,$2,$3,$4,$5)
			ON CONFLICT (id) DO UPDATE SET
			  sender = EXCLUDED.sender,
			  memo_id = EXCLUDED.memo_id,
			  payload = EXCLUDED.payload,
			  call_data_hash = EXCLUDED.call_data_hash
		`, e.Base.ID, e.Sender, e.MemoID, e.Payload, e.CallDataHash)
		return err

	case models.ERC8004Event:
		if err := upsertBase(ctx, tx, e.Base); err != nil {
			return err
		}
		_, err := tx.Exec(ctx, `
			INSERT INTO evt_erc8004 (id, registry, event_kind, agent_id, payload)
			VALUES ($1,$2,$3,$4,$5)
			ON CONFLICT (id) DO UPDATE SET
			  registry = EXCLUDED.registry,
			  event_kind = EXCLUDED.event_kind,
			  agent_id = EXCLUDED.agent_id,
			  payload = EXCLUDED.payload
		`, e.Base.ID, e.Registry, e.EventKind, nullIfEmpty(e.AgentID), e.Payload)
		return err

	case models.ERC8183Event:
		if err := upsertBase(ctx, tx, e.Base); err != nil {
			return err
		}
		_, err := tx.Exec(ctx, `
			INSERT INTO evt_erc8183 (id, job_id, event_kind, payload)
			VALUES ($1,$2,$3,$4)
			ON CONFLICT (id) DO UPDATE SET
			  job_id = EXCLUDED.job_id,
			  event_kind = EXCLUDED.event_kind,
			  payload = EXCLUDED.payload
		`, e.Base.ID, e.JobID, e.EventKind, e.Payload)
		return err

	case models.BaseEvent:
		return upsertBase(ctx, tx, e)

	default:
		return fmt.Errorf("unsupported event type %T", ev)
	}
}

func nullIfEmpty(s string) any {
	if s == "" {
		return nil
	}
	return s
}

func upsertBase(ctx context.Context, tx pgx.Tx, e models.BaseEvent) error {
	_, err := tx.Exec(ctx, `
		INSERT INTO evt_base (
		  id, chain_id, block_number, block_hash, tx_hash, log_index,
		  address, topic0, event_name, observed_at, raw_topics, raw_data, decoded, schema_version
		) VALUES (
		  $1,$2,$3,$4,$5,$6,
		  $7,$8,$9,$10,$11,$12,$13,$14
		)
		ON CONFLICT (chain_id, tx_hash, log_index) DO UPDATE SET
		  id = EXCLUDED.id,
		  block_number = EXCLUDED.block_number,
		  block_hash = EXCLUDED.block_hash,
		  address = EXCLUDED.address,
		  topic0 = EXCLUDED.topic0,
		  event_name = EXCLUDED.event_name,
		  raw_topics = EXCLUDED.raw_topics,
		  raw_data = EXCLUDED.raw_data,
		  decoded = EXCLUDED.decoded,
		  schema_version = EXCLUDED.schema_version
	`, e.ID, e.ChainID, e.BlockNumber, e.BlockHash, e.TxHash, e.LogIndex,
		e.Address, e.Topic0, e.EventName, e.ObservedAt, e.RawTopics, e.RawData, e.Decoded, e.SchemaVersion)
	return err
}

File: services/matcher/src/db/postgres.ts
import pg from "pg";
import type {
  Cursor,
  Erc8004Row,
  Erc8183Row,
  MemoRow,
  TransferRow,
} from "../types.js";

const { Pool } = pg;

export interface BaseEventRef {
  id: string;
  chain_id: string;
  block_number: string;
  tx_hash: string;
  log_index: number;
  event_name: string;
  address: string;
}

export class EventStore {
  readonly pool: pg.Pool;
  readonly chainId: string;
  readonly cursorId: string;

  constructor(databaseUrl: string, chainId: string) {
    this.pool = new Pool({ connectionString: databaseUrl });
    this.chainId = chainId;
    this.cursorId = `neo4j:${chainId}`;
  }

  async close() {
    await this.pool.end();
  }

  async ensureCursor(): Promise<void> {
    await this.pool.query(
      `INSERT INTO matcher_cursor (id, last_block, last_log_index)
       VALUES ($1, 0, 0)
       ON CONFLICT (id) DO NOTHING`,
      [this.cursorId],
    );
  }

  async getCursor(): Promise<Cursor> {
    await this.ensureCursor();
    const { rows } = await this.pool.query<{
      last_block: string;
      last_log_index: number;
    }>(`SELECT last_block, last_log_index FROM matcher_cursor WHERE id = $1`, [
      this.cursorId,
    ]);
    if (!rows[0]) return { lastBlock: 0, lastLogIndex: 0 };
    return {
      lastBlock: Number(rows[0].last_block),
      lastLogIndex: rows[0].last_log_index,
    };
  }

  async setCursor(cursor: Cursor): Promise<void> {
    await this.pool.query(
      `INSERT INTO matcher_cursor (id, last_block, last_log_index, updated_at)
       VALUES ($1, $2, $3, now())
       ON CONFLICT (id) DO UPDATE SET
         last_block = EXCLUDED.last_block,
         last_log_index = EXCLUDED.last_log_index,
         updated_at = now()`,
      [this.cursorId, cursor.lastBlock, cursor.lastLogIndex],
    );
  }

  /** Unified watermark stream for one chain — never skip events across families. */
  async fetchBasePage(cursor: Cursor, limit: number): Promise<BaseEventRef[]> {
    const { rows } = await this.pool.query(
      `SELECT id, chain_id::text, block_number::text, tx_hash, log_index, event_name, address
       FROM evt_base
       WHERE chain_id = $1
         AND (
           (block_number > $2)
           OR (block_number = $2 AND log_index > $3)
         )
       ORDER BY block_number, log_index
       LIMIT $4`,
      [this.chainId, cursor.lastBlock, cursor.lastLogIndex, limit],
    );
    return rows as BaseEventRef[];
  }

  async fetchTransfersByIds(ids: string[]): Promise<TransferRow[]> {
    if (ids.length === 0) return [];
    const { rows } = await this.pool.query(
      `SELECT b.id, b.chain_id::text, b.block_number::text, b.tx_hash, b.log_index,
              b.address, b.event_name, b.decoded,
              t.from_addr, t.to_addr, t.value::text, t.decimals, t.emitter_role
       FROM evt_transfer t
       JOIN evt_base b ON b.id = t.id
       WHERE t.id = ANY($1::text[])`,
      [ids],
    );
    return rows.map(normalizeDecoded) as TransferRow[];
  }

  async fetchMemosByIds(ids: string[]): Promise<MemoRow[]> {
    if (ids.length === 0) return [];
    const { rows } = await this.pool.query(
      `SELECT b.id, b.chain_id::text, b.block_number::text, b.tx_hash, b.log_index,
              b.address, b.event_name, b.decoded,
              m.sender, m.memo_id, m.payload, m.call_data_hash
       FROM evt_memo m
       JOIN evt_base b ON b.id = m.id
       WHERE m.id = ANY($1::text[])`,
      [ids],
    );
    return rows.map(normalizeDecoded) as MemoRow[];
  }

  async fetchMemosForTxs(txHashes: string[]): Promise<MemoRow[]> {
    if (txHashes.length === 0) return [];
    const { rows } = await this.pool.query(
      `SELECT b.id, b.chain_id::text, b.block_number::text, b.tx_hash, b.log_index,
              b.address, b.event_name, b.decoded,
              m.sender, m.memo_id, m.payload, m.call_data_hash
       FROM evt_memo m
       JOIN evt_base b ON b.id = m.id
       WHERE b.tx_hash = ANY($1::text[])`,
      [txHashes],
    );
    return rows.map(normalizeDecoded) as MemoRow[];
  }

  async fetchErc8183ByIds(ids: string[]): Promise<Erc8183Row[]> {
    if (ids.length === 0) return [];
    const { rows } = await this.pool.query(
      `SELECT b.id, b.chain_id::text, b.block_number::text, b.tx_hash, b.log_index,
              b.address, b.event_name, b.decoded,
              e.job_id, e.event_kind, e.payload
       FROM evt_erc8183 e
       JOIN evt_base b ON b.id = e.id
       WHERE e.id = ANY($1::text[])`,
      [ids],
    );
    return rows.map(normalizeDecoded) as Erc8183Row[];
  }

  async fetchErc8004ByIds(ids: string[]): Promise<Erc8004Row[]> {
    if (ids.length === 0) return [];
    const { rows } = await this.pool.query(
      `SELECT b.id, b.chain_id::text, b.block_number::text, b.tx_hash, b.log_index,
              b.address, b.event_name, b.decoded,
              e.registry, e.event_kind, e.agent_id, e.payload
       FROM evt_erc8004 e
       JOIN evt_base b ON b.id = e.id
       WHERE e.id = ANY($1::text[])`,
      [ids],
    );
    return rows.map(normalizeDecoded) as Erc8004Row[];
  }
}

function tryParseJson(value: unknown): unknown {
  if (value === null || value === undefined) return value;
  if (typeof value !== "string") return value;
  const s = value.trim();
  if (!s) return value;
  // Hex / opaque chain bytes (memo payload) are not JSON.
  if (s.startsWith("0x") || s.startsWith("0X")) return value;
  const head = s[0];
  if (head !== "{" && head !== "[" && head !== '"' && head !== "t" && head !== "f" && head !== "n" && !(head >= "0" && head <= "9") && head !== "-") {
    return value;
  }
  try {
    return JSON.parse(s);
  } catch {
    return value;
  }
}

function normalizeDecoded<T extends { decoded: unknown; payload?: unknown }>(
  row: T,
): T {
  const decoded = tryParseJson(row.decoded) ?? {};
  const payload =
    row.payload === undefined ? undefined : tryParseJson(row.payload);
  return { ...row, decoded, ...(payload !== undefined ? { payload } : {}) };
}
